### ISSUE TITLE
Wrap functions that require non-pointer struct arguments

### ISSUE THREAD

**Issue by RyanGlScott** [jwiegley/bindings-dsl#1]
The Haskell FFI cannot handle pure struct arguments in foreign declarations, only struct pointers. Unfortunately, I deal with several libraries with functions that expect pure structs as arguments. As a simplified example:

``` c
typedef struct {
    int x;
    int y;
} coord_t;

int coord_x(coord_t c) {
    return c.x;
}
```

There isn't any way for Haskell to call this directly. Instead, I'd first have to create another C file with a function that wraps `coord_x`:

``` c
int wrap_coord_x(coord_t *c) {
    return coord_x(*c);
}
```

This gets to be pretty tedious and mechanical, which made me wonder if this process could be automated via build-tools. It seems like `bindings-DSL` could be able to accomplish something like this via macros. Perhaps `bindings-DSL` could introduce notation such as

``` c
#cwrap coord_x , Ptr <coord_t> -> CInt
```

that could automatically generate

```
foreign import ccall "wrap_coord_x" c'coord_x
  :: Ptr C'coord_t -> CInt
```

where `wrap_coord_x` is created by `bindings-DSL`?

**Comment by RyanGlScott** [comment]
Update: this feature was [added to `c2hs`](https://github.com/haskell/c2hs/issues/117) a while back. The catch is that `Cabal` has difficulty finding generated `.c` files, so it has to be told explicitly where they are. The magnificent Ian Ross made a [pull request to `Cabal`](https://github.com/haskell/cabal/pull/2467) that informs it of preprocessor-generated `.c` files, but this probably won't be available until `Cabal-1.24`. Until then, I cooked up a [hacky `Setup.hs` script](https://github.com/haskell/c2hs/blob/8d88cf06c8ffd1f212af3b58a6eeafba97a13d86/custom-setup.hs) that achieves something similar.

`c2hs` laid out a proof-of-concept, and conceptually, it shouldn't be much more challenging to implement something like it in `bindings-dsl` using `hsc2hs`'s `#def` macro. The thing I'm not sure about is how `bindings-dsl` "parses" function signatures and distinguishes between integral types, pointers, and structs. @jwiegley, is there a good place to start in `bindings.dsl.h` to learn about that?

**Comment by RyanGlScott** [comment]
Closing stale issue.
